nx/lib/nx/random.ex

```diff
@@ -0,0 +1,390 @@
+  
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.uniform(key)
+      #Nx.Tensor<
+        f32[1]
+        [0.3037703037261963]
+      >
+  
+  If you need a new random number, you
+  can use `Nx.Random.split/2` to generate new subkeys:
+  
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.split(key)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [407137227, 1028236956],
+          [2469535296, 588450469]
+        ]
+      >
+  
+  ## Advanced
+  
+  Design and Context
+  
+  TLDR: Nx PRNG = Threefry counter PRNG + a functional array-oriented splitting model
+  
+  To summarize, among other requirements, the Nx PRNG aims to:
+  
+  1. ensure reproducibility,
+  
+  2. parallelize well, both in terms of vectorization (generating array values)
+     and multi-replica, multi-core computation. In particular it should not use
+     sequencing constraints between random function calls.
+  """
+
+  import Nx.Defn, only: [defn: 2, defnp: 2]
+  import Nx.Defn.Kernel, only: [assert_shape: 2, keyword!: 2]
+
+  defp assert_key(tensor) do
+    case Nx.shape(tensor) do
+      {2} ->
+        :ok
+
+      other ->
+        raise(
+          ArgumentError,
+          "expected key to have shape {2}, got tensor with shape #{Kernel.inspect(other)}"
+        )
+    end
+
+    type = Nx.type(tensor)
+
+    if not Nx.Type.integer?(type) do
+      raise ArgumentError,
+            "expected key with integer type, got key with type #{inspect(type)}"
+    end
+  end
+
+  @doc """
+  Create a pseudo-random number generator (PRNG) key given an integer seed.
+  
+  ## Examples
+  
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+  
+      iex> Nx.Random.key(999999999999)
+      #Nx.Tensor<
+        u32[2]
+        [232, 3567587327]
+      >
+  """
+  def key(seed) do
+    k1 = Nx.right_shift(seed, 32) |> Nx.reshape({1})
+    k2 = Nx.bitwise_and(seed, 0xFFFFFFFF) |> Nx.reshape({1})
+
+    Nx.concatenate([k1, k2])
+    |> Nx.reshape({2})
+    |> Nx.as_type({:u, 32})
+  end
+
+  @doc """
+  Splits a PRNG key into `num` new keys by adding a leading axis.
+  
+    ## Examples
+  
+      iex> key = Nx.Random.key(1701)
+      iex> Nx.Random.split(key, 2)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [56197195, 1801093307],
+          [961309823, 1704866707]
+        ]
+      >
+  
+      iex> key = Nx.Random.key(999999999999)
+      iex> Nx.Random.split(key, 4)
+      #Nx.Tensor<
+        u32[4][2]
+        [
+          [3959978897, 4079927650],
+          [3769699049, 3585271160],
+          [3182829676, 333122445],
+          [3185556048, 1258545461]
+        ]
+      >
+  """
+  def split(key, num \\ 2) when num > 1 do
+    assert_key(key)
+
+    impl(key, Nx.iota({num, 2}))
+  end
+
+  def fold_in(key, data) when is_integer(data) do
+    assert_key(key)
+
+    impl(key, key(data))
+  end
+
+  def random_bits(key, shape \\ {1}) do
+    assert_key(key)

```

Comment:
We left assert_key as it is to raise an appropriate error in case of wrong key type because there are mentioned functions that validate shape but I didn't find any that could validate type in defn.

Also, it's the first time I've heard of deftransform. It could use it's entry in documentation.

Reply:
It will be once 0.3 is released! It's a recent addition in an effort to make transforms easier to work with :)

Reply:
That's cool. I'll leave some memo to come back and change this once 0.3 hits the road.

Reply:
Oh, but since you're coding in Nx itself it already is available!

You can either generate the docs locally or search for the definition (in Nx.Defn or Nx.Defn.Kernel, can't recall which one :sweat-smile:)